src/transport/MessageEndpoint.cpp

```diff
@@ -0,0 +1,302 @@
+            }
+        } catch (zmq::error_t& e) {
+            logger->error("Error sending message: {}", e.what());
+            throw;
+        }
+    } else {
+        try {
+            auto res = this->socket->send(zmq::buffer(serialisedMsg, msgSize),
+                                          zmq::send_flags::none);
+            if (res != msgSize) {
+                throw std::runtime_error(
+                  fmt::format("Sent different bytes than expected (sent "
+                              "{}, expected {})",
+                              res.value_or(0),
+                              msgSize));
+            }
+        } catch (zmq::error_t& e) {
+            logger->error("Error sending message: {}", e.what());
+            throw;
+        }
+    }
+}
+
+// TODO - use optional size parameter
+Message MessageEndpoint::recv(int size)
+{
+    assert(tid == std::this_thread::get_id());
+    assert(this->socket != nullptr);
+    assert(size >= 0);
+
+    const auto& logger = faabric::util::getLogger();
+
+    // Pre-allocate buffer to avoid copying data
+    if (size > 0) {
+        Message msg(size);
+
+        try {
+            auto res = this->socket->recv(zmq::buffer(msg.udata(), msg.size()));
+            if (res.has_value() && (res->size != res->untruncated_size)) {
+                throw std::runtime_error(
+                  fmt::format("Received more bytes than buffer can hold. "
+                              "Received: {}, capacity {}",
+                              res->untruncated_size,
+                              res->size));
+            }
+        } catch (zmq::error_t& e) {
+            if (e.num() == ZMQ_ETERM) {
+                // Return empty message to signify termination
+                logger->trace("Shutting endpoint down after receiving ETERM");
+                return Message();
+            } else {
+                logger->error("Error receiving message: {}", e.what());
+                throw;
+            }
+        }
+
+        return msg;
+    }
+
+    // Allocate a message to receive data
+    zmq::message_t msg;
+    try {
+        auto res = this->socket->recv(msg);
+        if (!res.has_value()) {
+            throw std::runtime_error("Receive failed with EAGAIN");
+        }
+    } catch (zmq::error_t& e) {
+        if (e.num() == ZMQ_ETERM) {
+            // Return empty message to signify termination
+            logger->trace("Shutting endpoint down after receiving ETERM");
+            return Message();
+        } else {
+            logger->error("Error receiving message: {}", e.what());
+            throw;
+        }
+    }
+
+    // Copy the received message to a buffer whose scope we control
+    return Message(msg);
+}
+
+void MessageEndpoint::close(bool bind)
+{
+    if (this->socket != nullptr) {
+        const auto& logger = faabric::util::getLogger();
+
+        if (tid != std::this_thread::get_id()) {
+            logger->error("Closing socket from a different thread");
```

Comment:
Having an error line with no exception or failure seems strange. Either it's something we expect, which isn't necessarily a failure but should be obvious when investigating issues (in which case it would be a `warn`), or it's a failure, in which case we'd put more information in this `error` log (e.g. which thread IDs it is), then throw an exception.